dsc/reports/report.py

```diff
@@ -0,0 +1,184 @@
+from jinja2 import Environment, FileSystemLoader, Template, select_autoescape
+
+from dsc.config import Config
+from dsc.db.models import ItemSubmissionStatus
+from dsc.item_submission import ItemSubmission
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from dsc import workflows
+
+logger = logging.getLogger(__name__)
+CONFIG = Config()
+
+
+class Report(ABC):
+    """Base class for all DSC reporting."""
+
+    fields: ClassVar[list[str]] = [
+        "batch_id",
+        "item_identifier",
+        "source_system_identifier",
+        "status",
+        "status_details",
+        "dspace_handle",
+        "ingest_date",
+    ]
+
+    def __init__(self, workflow_name: str, batch_id: str):
+        self.workflow_name = workflow_name
+        self.batch_id = batch_id
+        self.client = boto3.client("dynamodb")
+        self.report_date = datetime.datetime.now(tz=datetime.UTC).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+        # configure environment for loading jinja templates
+        self.jinja_env = Environment(
+            loader=FileSystemLoader("dsc/reports/templates/"),
+            autoescape=select_autoescape(),
+        )
+
+        # cache list of item submissions
+        self._item_submissions: list[dict] | None = None
+
+    @property
+    @abstractmethod
+    def subject(self) -> str:
+        """Subject heading for report used in email."""
+
+    @property
+    @abstractmethod
+    def summary_template(self) -> Template:
+        """Jinja template for report summary."""
+
+    @abstractmethod
+    def generate_summary(self) -> str:
+        """Render summary from report template.
+
+        The generated summary will be used as the email message body.
+        """
+
+    @property
+    def batch_item_submissions(self) -> list[dict]:
+        """Report data retrieved from DynamoDB.
+
+        An effect of running Workflow.create_batch is the writing of item submissions
+        to DynamoDB. Therefore, this report reads the following data from DynamoDB:
+
+        - All item submissions for a given batch
+        """
+        if not self._item_submissions:
+            self._item_submissions = [
+                item_submission.asdict_subset(fields=self.fields)
+                for item_submission in ItemSubmission.get_batch(self.batch_id)
+            ]
+        return self._item_submissions
+
+    @classmethod
+    def from_workflow(cls, workflow: workflows.Workflow) -> Report:
+        """Create instance of Report using dsc.workflows.Workflow."""
+        return cls(
+            workflow_name=workflow.workflow_name,
+            batch_id=workflow.batch_id,
+        )
+
+    def write_item_submissions_csv(
+        self, output_file: StringIO | str | Path
+    ) -> StringIO | str | Path:
+        """Write report data to either a CSV file or in-memory string buffer."""
+        if isinstance(output_file, StringIO):
+            logger.debug("Writing data to CSV buffer")
+        else:
```

Comment:
Are we ever writing to an output file? I only see a call to a buffer in `base/__init__.py` line 582, do we plan to use the ability to write to a file? If not, we could simplify this method

Reply:
The option to write to an output file may be useful for local testing! :)

Reply:
Fair enough!